### PR TITLE
Create transitionState function

### PR DESCRIPTION
**Pull request by fractalwrench**
## Goal

Refactors SessionOrchestratorImpl to create a transitionState function that models the logic required for each state transition. This makes it less likely we will forget to perform a step if we ever need to add an extra transition in future, and makes the code easier to read.

## Testing

Relied on existing test coverage.

### CODE REVIEW

File: embrace-android-sdk/src/main/java/io/embrace/android/embracesdk/session/orchestrator/SessionOrchestratorImpl.kt [hunk 1]
```diff
@@ -51,117 +56,130 @@
...
-                        "many sessions"
-                )
-                return
+        val timestamp = clock.now()
+
+        transitionState(
+            description = "endManual",
+            timestamp = timestamp,
+            clearUserInfo = clearUserInfo,
+            endProcessState = State.FOREGROUND,
```
Comment (bidetofevil) [inline embrace-android-sdk/src/main/java/io/embrace/android/embracesdk/session/orchestrator/SessionOrchestratorImpl.kt:122]: Could this be triggered in the background? That is, can we end background activities manually? 

I'm wondering if this is something we want to derive given the current state rather than stipulate before we transition, as doing this manually seems unrelated to background vs foreground.

File: embrace-android-sdk/src/main/java/io/embrace/android/embracesdk/session/orchestrator/SessionOrchestratorImpl.kt [hunk 2]
```diff
@@ -171,6 +189,66 @@
...
+        endProcessState: State,
+        oldSessionAction: (initial: Session) -> Unit,
+        newSessionAction: () -> Session?,
+        earlyTerminationCondition: () -> Boolean = { false },
+        clearUserInfo: Boolean = false,
+    ) {
+        // supplied business logic says that we can't perform a transition yet.
+        // exit early & retain the current state instead.
+        if (earlyTerminationCondition()) {
+            return
```
Comment (bidetofevil) [inline embrace-android-sdk/src/main/java/io/embrace/android/embracesdk/session/orchestrator/SessionOrchestratorImpl.kt:221]: Is there utility in returning whether this succeeded or not? Might be an interesting case to log an internal error/warning for if this is unexpected

**Review by bidetofevil — APPROVED** [review]
I have a couple questions about edge cases but looks good otherwise. Feel free to merge away if the questions don't spark anything that you want to change now.

File: embrace-android-sdk/src/main/java/io/embrace/android/embracesdk/session/orchestrator/SessionOrchestratorImpl.kt (hunk 1 above)
Comment (fractalwrench) [reply embrace-android-sdk/src/main/java/io/embrace/android/embracesdk/session/orchestrator/SessionOrchestratorImpl.kt:122]: I don't think we can derive this from current state (at least without passing in extra information during `createInitialEnvelope`).

File: embrace-android-sdk/src/main/java/io/embrace/android/embracesdk/session/orchestrator/SessionOrchestratorImpl.kt (hunk 2 above)
Comment (fractalwrench) [reply embrace-android-sdk/src/main/java/io/embrace/android/embracesdk/session/orchestrator/SessionOrchestratorImpl.kt:221]: I've added some extra logging for this in this specific class. There are already some existing internal errors logged from ProcessStateService if an unbalanced call happens.

**Comment by fractalwrench** [comment]
Tagging for review again as I needed a rebase & a few fixes for failing tests.

**Review by bidetofevil — APPROVED** [review]
Looks good
